src/backend/distributed/commands/multi_copy.c

```diff
@@ -2923,3 +3070,380 @@ CopyGetAttnums(TupleDesc tupDesc, Relation rel, List *attnamelist)
+{
+	List *connectionStateList = NIL;
+	HASH_SEQ_STATUS status;
+	CopyConnectionState *connectionState = NULL;
+
+	if (connectionStateHash == NULL)
+	{
+		return NIL;
+	}
+
+	hash_seq_init(&status, connectionStateHash);
+
+	connectionState = (CopyConnectionState *) hash_seq_search(&status);
+	while (connectionState != NULL)
+	{
+		connectionStateList = lappend(connectionStateList, connectionState);
+
+		connectionState = (CopyConnectionState *) hash_seq_search(&status);
+	}
+
+	return connectionStateList;
+}
+
+
+/*
+ * GetShardState finds existing CopyShardState for a shard in the provided
+ * hash. If not found, then a new shard state is returned with all related
+ * CopyPlacementStates initialized.
+ */
+static CopyShardState *
+GetShardState(uint64 shardId, HTAB *shardStateHash,
+			  HTAB *connectionStateHash, bool stopOnFailure, bool *found)
+{
+	CopyShardState *shardState = NULL;
+
+	shardState = (CopyShardState *) hash_search(shardStateHash, &shardId,
+												HASH_ENTER, found);
+	if (!*found)
+	{
+		InitializeCopyShardState(shardState, connectionStateHash,
+								 shardId, stopOnFailure);
+	}
+
+	return shardState;
+}
+
+
+/*
+ * InitializeCopyShardState initializes the given shardState. It finds all
+ * placements for the given shardId, assignes connections to them, and
+ * adds them to shardState->placementStateList.
+ */
+static void
+InitializeCopyShardState(CopyShardState *shardState,
+						 HTAB *connectionStateHash, uint64 shardId,
+						 bool stopOnFailure)
+{
+	List *finalizedPlacementList = NIL;
+	ListCell *placementCell = NULL;
+	bool failedPlacementCount = 0;
+
+	MemoryContext localContext =
+		AllocSetContextCreateExtended(CurrentMemoryContext,
+									  "InitializeCopyShardState",
+									  ALLOCSET_DEFAULT_MINSIZE,
+									  ALLOCSET_DEFAULT_INITSIZE,
+									  ALLOCSET_DEFAULT_MAXSIZE);
+
+	/* release finalized placement list at the end of this function */
+	MemoryContext oldContext = MemoryContextSwitchTo(localContext);
+
+	finalizedPlacementList = MasterShardPlacementList(shardId);
+
+	MemoryContextSwitchTo(oldContext);
+
+	shardState->shardId = shardId;
+	shardState->placementStateList = NIL;
+
+	foreach(placementCell, finalizedPlacementList)
+	{
+		ShardPlacement *placement = (ShardPlacement *) lfirst(placementCell);
+		CopyConnectionState *connectionState = NULL;
+		CopyPlacementState *placementState = NULL;
+		
+		MultiConnection *connection =
+			CopyGetPlacementConnection(placement, connectionStateHash, stopOnFailure);
+		if (connection == NULL)
+		{
+			failedPlacementCount++;
+			continue;
+		}
+
+		connectionState = GetConnectionState(connectionStateHash, connection);
+		placementState = palloc0(sizeof(CopyPlacementState));
+		placementState->shardState = shardState;
+		placementState->data = makeStringInfo();
+		placementState->connectionState = connectionState;
+		dlist_push_head(&connectionState->bufferedPlacementList,
+						&placementState->bufferedPlacementNode);
```

Comment:
When there is not yet an active placement, it's slightly confusing to me that we add the placement to `bufferedPlacementList` only to remove it right after.

Could we make the logic more explicit by adding something like:
```c
if (connectionState->activePlacementState == NULL)
{
     connectionState->activePlacementState = placementState;
}
else
{
    dlist_push_head(&connectionState->bufferedPlacementList, &placementState->bufferedPlacementNode);
}
```